packages/babel-parser/src/parser/statement.ts

```diff
@@ -346,15 +346,26 @@ export default abstract class StatementParser extends ExpressionParser {
     );
   }
 
-  startsUsingForOf(): boolean {
-    const { type, containsEsc } = this.lookahead();
+  allowsForUsing(): boolean {
+    const { type, containsEsc, start } = this.lookahead();
     if (type === tt._of && !containsEsc) {
-      // `using of` must start a for-lhs-of statement
-      return false;
-    } else if (tokenIsIdentifier(type) && !this.hasFollowingLineBreak()) {
+      // `for( using of` must start either a for-lhs-of statement
+      // or a for lexical declaration
+      const nextCharAfterOf = this.input.charCodeAt(
+        this.nextTokenStartSince(start + 2),
+      );
+      if (
+        nextCharAfterOf !== charCodes.equalsTo &&
+        nextCharAfterOf !== charCodes.colon
```

Comment:
We could also add here `charCodes.semicolon`, to give the good error message for `for (using x;;);`